src/main/java/io/vertx/ext/asyncsql/impl/pool/AsyncConnectionPool.java

```diff
@@ -0,0 +1,126 @@
+package io.vertx.ext.asyncsql.impl.pool;
+
+import com.github.mauricio.async.db.Configuration;
+import com.github.mauricio.async.db.Connection;
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Future;
+import io.vertx.core.Handler;
+import io.vertx.core.Vertx;
+import io.vertx.core.logging.Logger;
+import io.vertx.core.logging.LoggerFactory;
+import io.vertx.ext.asyncsql.impl.ScalaUtils;
+import io.vertx.ext.asyncsql.impl.VertxExecutionContext;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public abstract class AsyncConnectionPool {
+
+  private final int maxPoolSize;
+
+  private static final Logger logger = LoggerFactory.getLogger(AsyncConnectionPool.class);
+  protected final Configuration configuration;
+  protected final Vertx vertx;
+
+  private int poolSize = 0;
+  private final List<Connection> availableConnections = new ArrayList<>();
+  private final List<Handler<AsyncResult<Connection>>> waiters = new ArrayList<>();
+
+  public AsyncConnectionPool(Vertx vertx, int maxPoolSize, Configuration configuration) {
+    this.vertx = vertx;
+    this.maxPoolSize = maxPoolSize;
+    this.configuration = configuration;
+    //TODO would it make sense to pass the execution context here ?
+  }
+
+  protected abstract Connection create();
+
+  private synchronized void createConnection(Handler<AsyncResult<Connection>> handler) {
+    poolSize += 1;
+    try {
+      Connection connection = create();
+      connection
+          .connect().onComplete(ScalaUtils.toFunction1(handler),
+          VertxExecutionContext.create(vertx));
```

Comment:
See comment above, it should be possible to keep a reference to an execution context and not recreate one on each `createConnection` call.